### PR TITLE
gomod: only support one gomod in tools and integrations respectively

### PR DESCRIPTION
**Pull request by HuSharp**
### What problem does this PR solve?

Issue Number: Ref #7300

### What is changed and how does it work?

```commit-message
nowadays we have too many gomods which are redundant

remove redundant gomods, support only one go mod in integrations/tools/client/server respectively
```

- for tools:
```
$ rm -rf bin && make build && make tools && ls bin                                                                                                                                                                     
        
./scripts/embed-dashboard-ui.sh
+ Clean up existing asset file
+ Fetch TiDB Dashboard Go module
  - TiDB Dashboard directory: /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/pingcap/tidb-dashboard@v0.0.0-20231218095437-aa621ed4de2c
+ Create download cache directory: /data/nvme0n1/husharp/proj/pd/.dashboard_download_cache
+ Discover TiDB Dashboard release version
  - TiDB Dashboard release version: 2023.12.18.1
+ Check whether pre-built assets are available
  - Cached archive exists: /data/nvme0n1/husharp/proj/pd/.dashboard_download_cache/embedded-assets-golang-2023.12.18.1.zip
+ Unpack embedded asset from archive
Archive:  /data/nvme0n1/husharp/proj/pd/.dashboard_download_cache/embedded-assets-golang-2023.12.18.1.zip
  inflating: embedded_assets_handler.go  
  - Unpacked /data/nvme0n1/husharp/proj/pd/pkg/dashboard/uiserver/embedded_assets_handler.go
GOEXPERIMENT= CGO_ENABLED=1 go build  -gcflags '' -ldflags '-X "github.com/tikv/pd/pkg/versioninfo.PDReleaseVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/tikv/pd/pkg/versioninfo.PDBuildTS=2023-12-29 09:08:29" -X "github.com/tikv/pd/pkg/versioninfo.PDGitHash=c5f2cb6002d533ade8df4e6c6c021b163d6491bf" -X "github.com/tikv/pd/pkg/versioninfo.PDGitBranch=replace_ctl_test_gomod" -X "github.com/tikv/pd/pkg/versioninfo.PDEdition=Community" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.InternalVersion=2023.12.18.1" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.Standalone=No" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.PDVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildTime=2023-12-29 09:08:29" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildGitHash=aa621ed4de2c"' -tags "" -o /data/nvme0n1/husharp/proj/pd/bin/pd-server cmd/pd-server/main.go
cd tools && GOEXPERIMENT= CGO_ENABLED=0 go build -gcflags '' -ldflags '-X "github.com/tikv/pd/pkg/versioninfo.PDReleaseVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/tikv/pd/pkg/versioninfo.PDBuildTS=2023-12-29 09:08:36" -X "github.com/tikv/pd/pkg/versioninfo.PDGitHash=c5f2cb6002d533ade8df4e6c6c021b163d6491bf" -X "github.com/tikv/pd/pkg/versioninfo.PDGitBranch=replace_ctl_test_gomod" -X "github.com/tikv/pd/pkg/versioninfo.PDEdition=Community" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.InternalVersion=2023.12.18.1" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.Standalone=No" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.PDVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildTime=2023-12-29 09:08:36" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildGitHash=aa621ed4de2c"' -o /data/nvme0n1/husharp/proj/pd/bin/pd-ctl pd-ctl/main.go
cd tools && GOEXPERIMENT= CGO_ENABLED=0 go build -gcflags '' -ldflags '-X "github.com/tikv/pd/pkg/versioninfo.PDReleaseVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/tikv/pd/pkg/versioninfo.PDBuildTS=2023-12-29 09:08:38" -X "github.com/tikv/pd/pkg/versioninfo.PDGitHash=c5f2cb6002d533ade8df4e6c6c021b163d6491bf" -X "github.com/tikv/pd/pkg/versioninfo.PDGitBranch=replace_ctl_test_gomod" -X "github.com/tikv/pd/pkg/versioninfo.PDEdition=Community" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.InternalVersion=2023.12.18.1" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.Standalone=No" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.PDVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildTime=2023-12-29 09:08:38" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildGitHash=aa621ed4de2c"' -o /data/nvme0n1/husharp/proj/pd/bin/pd-recover pd-recover/main.go
cd tools && CGO_ENABLED=0 go build -o /data/nvme0n1/husharp/proj/pd/bin/pd-tso-bench pd-tso-bench/main.go
cd tools && CGO_ENABLED=0 go build -gcflags '' -ldflags '-X "github.com/tikv/pd/pkg/versioninfo.PDReleaseVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/tikv/pd/pkg/versioninfo.PDBuildTS=2023-12-29 09:08:40" -X "github.com/tikv/pd/pkg/versioninfo.PDGitHash=c5f2cb6002d533ade8df4e6c6c021b163d6491bf" -X "github.com/tikv/pd/pkg/versioninfo.PDGitBranch=replace_ctl_test_gomod" -X "github.com/tikv/pd/pkg/versioninfo.PDEdition=Community" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.InternalVersion=2023.12.18.1" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.Standalone=No" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.PDVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildTime=2023-12-29 09:08:40" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildGitHash=aa621ed4de2c"' -o /data/nvme0n1/husharp/proj/pd/bin/pd-heartbeat-bench pd-heartbeat-bench/main.go
cd tools && CGO_ENABLED=0 go build -gcflags '' -ldflags '-X "github.com/tikv/pd/pkg/versioninfo.PDReleaseVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/tikv/pd/pkg/versioninfo.PDBuildTS=2023-12-29 09:08:42" -X "github.com/tikv/pd/pkg/versioninfo.PDGitHash=c5f2cb6002d533ade8df4e6c6c021b163d6491bf" -X "github.com/tikv/pd/pkg/versioninfo.PDGitBranch=replace_ctl_test_gomod" -X "github.com/tikv/pd/pkg/versioninfo.PDEdition=Community" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.InternalVersion=2023.12.18.1" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.Standalone=No" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.PDVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildTime=2023-12-29 09:08:42" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildGitHash=aa621ed4de2c"' -o /data/nvme0n1/husharp/proj/pd/bin/regions-dump regions-dump/main.go
cd tools && CGO_ENABLED=0 go build -gcflags '' -ldflags '-X "github.com/tikv/pd/pkg/versioninfo.PDReleaseVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/tikv/pd/pkg/versioninfo.PDBuildTS=2023-12-29 09:08:43" -X "github.com/tikv/pd/pkg/versioninfo.PDGitHash=c5f2cb6002d533ade8df4e6c6c021b163d6491bf" -X "github.com/tikv/pd/pkg/versioninfo.PDGitBranch=replace_ctl_test_gomod" -X "github.com/tikv/pd/pkg/versioninfo.PDEdition=Community" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.InternalVersion=2023.12.18.1" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.Standalone=No" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.PDVersion=v6.2.0-alpha-1161-gc5f2cb600" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildTime=2023-12-29 09:08:43" -X "github.com/pingcap/tidb-dashboard/pkg/utils/version.BuildGitHash=aa621ed4de2c"' -o /data/nvme0n1/husharp/proj/pd/bin/stores-dump stores-dump/main.go
cd tools && CGO_ENABLED=0 go build -o /data/nvme0n1/husharp/proj/pd/bin/pd-api-bench pd-api-bench/main.go
pd-api-bench  pd-ctl  pd-heartbeat-bench  pd-recover  pd-server  pd-tso-bench  regions-dump  stores-dump
```

- for integrations test can be checked in ci action
client: https://github.com/tikv/pd/actions/runs/7354548836/job/20021908203?pr=7635
tso: https://github.com/tikv/pd/actions/runs/7354548836/job/20021908307?pr=7635
mcs: https://github.com/tikv/pd/actions/runs/7354548836/job/20021908434?pr=7635

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None.
```

### CODE REVIEW

File: tools/pd-ctl/go.mod [hunk 1]
```diff
@@ -0,0 +1,139 @@
+module github.com/tikv/pd/tools/pd-ctl
```
Comment (CabinfeverB) [inline tools/pd-ctl/go.mod:1]: should modify makefile

Comment (HuSharp) [reply tools/pd-ctl/go.mod:1]: done

File: tests/dashboard/service_test.go [hunk 2]
```diff
@@ -150,15 +147,28 @@ func (suite *dashboardTestSuite) testDashboard(re *require.Assertions, internalP
...
 	re.NoError(err)
+	req, _ := http.NewRequest(http.MethodPost, leaderAddr+"/pd/api/v1/config", bytes.NewBuffer(data))
+	resp, err := suite.httpClient.Do(req)
+	re.NoError(err)
+	resp.Body.Close()
 	suite.checkServiceIsStarted(re, internalProxy, servers, leader)
 	re.Equal(dashboardAddress2, leader.GetServer().GetPersistOptions().GetDashboardAddress())
 
 	// pd-ctl set stop
-	args = []string{"-u", leaderAddr, "config", "set", "dashboard-address", "none"}
```
Comment (rleungx) [inline tests/dashboard/service_test.go:160]: Shall we also test it in pd-ctl?

File: tools/pd-ctl/tests/keyspace/keyspace_test.go [hunk 3]
```diff
@@ -134,24 +134,24 @@ func TestKeyspaceGroupUninitialized(t *testing.T) {
...
 	ctx     context.Context
 	cancel  context.CancelFunc
-	cluster *tests.TestCluster
+	cluster *pdTests.TestCluster
 	pdAddr  string
 }
 
-func TestKeyspaceTestSuite(t *testing.T) {
-	suite.Run(t, new(keyspaceTestSuite))
+func TestKeyspacepdTestsuite(t *testing.T) {
```
Comment (rleungx) [inline tools/pd-ctl/tests/keyspace/keyspace_test.go:145]: pd?

File: tests/dashboard/service_test.go (hunk 2 above)
Comment (HuSharp) [reply tests/dashboard/service_test.go:160]: Yes we have [`config` test](tools/pd-ctl/tests/config/config_test.go), I'm guessing it was written that way for convenience which introduced unnecessary dependency

File: tools/pd-ctl/tests/keyspace/keyspace_test.go (hunk 3 above)
Comment (HuSharp) [reply tools/pd-ctl/tests/keyspace/keyspace_test.go:145]: sorry, Lazy replacement 0 0

File: scripts/update-dashboard.sh [hunk 4]
```diff
@@ -24,9 +24,5 @@ go mod tidy
...
 go mod tidy
 
-cd tests/integrations/client
-go mod tidy
-cd ../mcs
-go mod tidy
-cd ../tso
-go mod tidy
+cd ./tests/integrations
+go mod tidy
```
Comment (rleungx) [inline scripts/update-dashboard.sh:28]: leave an empty line

File: pd.code-workspace [hunk 5]
```diff
@@ -25,12 +25,8 @@
 			"path": "tests/integrations/realcluster"
 		},
 		{
-			"name": "pd-tso-bench",
-			"path": "tools/pd-tso-bench"
-		},
-		{
-			"name": "pd-api-bench",
-			"path": "tools/pd-api-bench"
+			"name": "tools",
```
Comment (rleungx) [inline pd.code-workspace:28]: Shall we also merge integrations?

Comment (HuSharp) [reply pd.code-workspace]: fixed

File: tools/pd-ctl/README.md [hunk 6]
```diff
@@ -4,7 +4,7 @@
 
 ## Build
 
-1. [Go](https://golang.org/) Version 1.16 or later
+1. [Go](https://golang.org/) Version 1.21 or later
```
Comment (CabinfeverB) [inline tools/pd-ctl/README.md:7]: also in pd-recover README?

Comment (HuSharp) [reply tools/pd-ctl/README.md:7]: good catch, fixed all doc

File: tools/go.mod [hunk 7]
```diff
@@ -1,31 +1,40 @@
...
+replace google.golang.org/grpc v1.59.0 => google.golang.org/grpc v1.26.0
+
 replace (
-	github.com/tikv/pd => ../../../
-	github.com/tikv/pd/client => ../../../client
+	github.com/tikv/pd => ../
+	github.com/tikv/pd/client => ../client
 )
 
-// reset grpc and protobuf deps in order to import client and server at the same time
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply tools/go.mod:10]: added

**Review by CabinfeverB — COMMENTED** [review]
mostly LGTM

**Review by nolouch — APPROVED** [review]
lgtm! great works.

**Comment by CabinfeverB** [comment]
Pls add some test results.

**Comment by HuSharp** [comment]
> Pls add some test results.

added in pr description

**Review by CabinfeverB — APPROVED** [review]

**Pull request merged**